src/Service/ForecastReportService.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+namespace App\Service;
+
+use App\Model\Reports\ForecastReportData;
+use App\Model\Reports\ForecastReportIssueData;
+use App\Model\Reports\ForecastReportIssueVersionData;
+use App\Model\Reports\ForecastReportProjectData;
+use App\Model\Reports\ForecastReportWorklogData;
+use App\Repository\WorkerRepository;
+use App\Repository\WorklogRepository;
+
+class ForecastReportService
+{
+    public function __construct(
+        private readonly WorklogRepository $worklogRepository,
+        private readonly WorkerRepository $workerRepository,
+    ) {
+    }
+
+    /**
+     * Get forecast report data based on given date range.
+     *
+     * @param \DateTimeInterface $fromDate The start date of the period
+     * @param \DateTimeInterface $toDate The end date of the period
+     *
+     * @return ForecastReportData The forecast report data
+     *
+     * @throws \Exception
+     */
+    public function getForecastReport(\DateTimeInterface $fromDate, \DateTimeInterface $toDate): ForecastReportData
+    {
+        // Get all worklogs attached to an invoice for the period
+        $invoiceAttachedWorklogs = $this->worklogRepository->getWorklogsAttachedToInvoiceInDateRange($fromDate, $toDate);
+
+        // Create an new instance of ForecastReportData
+        $forecastReportData = new ForecastReportData();
+
+        // Loop through each worklog
+        foreach ($invoiceAttachedWorklogs as $worklog) {
+            $projectId = $worklog->getProject()->getId();
+
+            if (!$projectId) {
+                throw new \Exception('Project id is null');
+            }
+            $projectName = $worklog->getProject()?->getName() ?? '[no project name]';
+
+            // If the project isn't already in the forecast, add it
+            if (!isset($forecastReportData->projects[$projectId])) {
+                $forecastReportData->projects[$projectId] = new ForecastReportProjectData($projectId);
+            }
+
+            if (is_object($forecastReportData->projects[$projectId])) {
```

Comment:
Why is this if needed?

Reply:
Primarily because psalm i worried that $forecastReportData->projects[$projectId] is possible null. I rewrote it to check if it is an instance of ForecastReportProjectData, to make it feel a bit more cohesive.